[IMP] doc: add information on the new populate feature. (#50596)
odoo/odoo: doc/reference/testing.rst

:option:`--screenshots <odoo-bin --screenshots>` and :option:`--screencasts <odoo-bin --screencasts>`

Performance Testing
===================

Query counts
------------

One of the ways to test performance is to measure database queries.
Manually, this can be tested with the `--log-sql` CLI parameter.

> If you want to fix the maximum number of queries for an operation,

Reviewer: ```suggestion
If you want to establish the maximum number of queries for an operation,
```
`fix` is a bit ambigous (correct vs stable)

you can use the :meth:`~odoo.tests.common.BaseCase.assertQueryCount` method,

> integrated in Odoo test classes.

Reviewer 2: RST line break <> HTML line break

.. code-block:: python

    with self.assertQueryCount(11):
        do_something()

.. _reference/testing/populate:

Database population
-------------------

Odoo CLI offers a :ref:`database population<reference/cmdline/populate>` feature.

.. code-block:: console

    odoo-bin populate

Instead of the tedious manual, or programmatic, specification of test data,
one can use this feature to fill a database on demand with the desired number of test data.
This can be used to detect diverse bugs or performance issues in tested flows.

> To enable this feature for a given model, the following methods|attributes must be defined.

Reviewer: ```suggestion
To enable this feature for a given model, the following methods and attributes must be defined.
```
We are not in a regex ;-)

> They all have a default in the Base model, but you have to define at least the method

Reviewer 2: ```suggestion
They all have a default in ``BaseModel``, but you have to define at least the method
```
Not sure about this... But I find that "the Base model" doesn't immediately refer to odoo.models.BaseModel

Reviewer: Why not indicating the default directly in the parameters instead of this sentence (and required for `_populate_factories`)?

:meth:`~odoo.models._populate_factories` to support model population.

.. _reference/testing/populate/methods:

.. currentmodule:: odoo.models

.. autoattribute:: Model._populate_sizes
.. autoattribute:: Model._populate_dependencies
.. automethod:: Model._populate
.. automethod:: Model._populate_factories

Example model
^^^^^^^^^^^^^

.. code-block:: python

    from odoo.tools import populate

    class CustomModel(models.Model)
        _inherit = "custom.some_model"
        _populate_sizes = {"small": 100, "medium": 2000, "large": 10000}
        _populate_dependencies = ["custom.some_other_model"]

        def _populate_factories(self):

>             some_other_ids = self.env.registry.populated_models["custom.some_other_model"]

Reviewer: add comment to explain what it does

            def get_some_other_id(random=None, **kwargs):
                return populate.randomize(some_other_ids)

            return [
                ("field_1", populate.randomize([value1, value2, value3]),
                ("field_2", populate.randomize([value_a, value_b], [0.5, 0.5]),

>                 ("some_other_id", populate.compute(get_some_other_id)),

Reviewer: There is some arguments on your `get_some_other_id` method. How can I use it?
